feature: Added post API

dominic-chang opened the pull request:
Created a page where clients can create new projects. Created a client side API that posts new projects to the database 'Project' table  and uplaods image to AWS S3 bucket specified in the .env file. The S3 object key is specifie by the project id.

dominic-chang committed:
feature: Added post API

Created a page where clients can create new projects. Created a client side API that posts new projects to the database 'Project' table  and uplaods image to AWS S3 bucket specified in the .env file. The S3 object key is specifie by the project id.

tevyt commented on package.json:
```diff
@@ -17,13 +17,17 @@
     "autoprefixer": "10.4.15",
+    "aws-sdk": "^2.1459.0",
+    "axios": "^1.5.0",
+    "fs": "^0.0.1-security",
```
fs is a node API, you shouldn't need to install this library.

tevyt commented on src/app/projects/new/page.tsx:
```diff
@@ -0,0 +1,124 @@
+        console.log(err);
+      })
+  };
+  let $imagePreview = null;
```
Leading `$` is not typically used for JavaScript variable names. Looks like a php thing.

tevyt commented on package.json:
```diff
@@ -17,13 +17,17 @@
     "@types/react-dom": "18.2.7",
     "autoprefixer": "10.4.15",
+    "aws-sdk": "^2.1459.0",
+    "axios": "^1.5.0",
```
Don't think we really needed Axios here, fetch was probably enough.

tevyt commented on package.json:
```diff
@@ -17,13 +17,17 @@
     "@types/react": "18.2.21",
     "@types/react-dom": "18.2.7",
     "autoprefixer": "10.4.15",
+    "aws-sdk": "^2.1459.0",
```
Unused dependency, we should uninstall this.

tevyt commented on src/app/projects/api/route.ts:
```diff
@@ -8,3 +10,49 @@ export async function GET() {
+
+  const projectService = new ProjectService();
+  const res = await projectService.postProject(prismaCreationRequest)
+    .then(async (res) => {
```
Don't need the `then` callback within an async function.

```js
const res = await projectService.postProject(prismaCreationClient);

console.log(res.id);

...
```

The `async` keyword creates an abstraction over promises, within the scope of an async function you can work with promises as if they were running synchronously. See [example](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/async_function)

tevyt commented on src/app/projects/api/route.ts:
```diff
@@ -8,3 +10,49 @@ export async function GET() {
+      };
+
+      return await client.send(new PutObjectCommand(putObjectParams))
+        .then(async (resAWS) => {
```
Same here as above, don't mix the `then` callback and `async` functions.

tevyt commented on src/app/projects/api/route.ts:
```diff
@@ -8,3 +10,49 @@ export async function GET() {
+        }
+        )
+        .catch((error) => {
+          console.log("Failed to Upload Image");
```
Use appropriate log level. `console.log` produces logs at the `INFO` level, we want to be able to separate these from errors so that we can easily identify errors when debugging. `console.error` produces error logs

https://www.atatus.com/blog/javascript-logging-basic-tips/
https://nodejs.org/api/console.html#consoleerrordata-args

tevyt commented on src/app/projects/api/route.ts:
```diff
@@ -8,3 +10,49 @@ export async function GET() {
+          console.log(`Uploaded logo to S3 with id: ${res.id}`);
+        }
+        )
+        .catch((error) => {
```
For error handling within an async function you can use exception handling.

```js
async function upload(){
  try{
    // asynchrous code
  }catch(error){
     console.error(error);
 }
```

tevyt commented on src/app/projects/api/route.ts:
```diff
@@ -8,3 +10,49 @@ export async function GET() {
+    })
+    .catch((error) => { console.log(error); });
+
+  return NextResponse.json({ res });
```
Be intentional at names, especially at boundary points. Calling `NextResponse.json({ res })` is going to result in a JSON response with res being the only key.

IE
```json
{
  "res" : {
     "name": "Cool project",
      ...
   }
}
```

While this isn't the worst, it is important to pay attention to contract points like this. APIs often have many consumers and it is important that any keys returned by them are easily understood.

I would instead simply return an object containing a project `NextResponse.json(res)` so that the project object stored at res is returned, rather than an adding an external wrapping object.

tevyt reviewed (changes requested):
Few issues we should clean up before merging. I'll handle the tests, just take a look at the suggestions on the MR.

tevyt reviewed (changes requested):
Since the dependencies have changed, we need the updated package-lock.json

dominic-chang committed:
fix: Cleaned up pull request

Removed unused dependencies. Replaced long promise chains with try catch
block. Added documentation for project configuration.

dominic-chang committed:
Merge branch 'main' into implement-project-creation

dominic-chang committed:
Fix bug

* Fixed bug that prevented logo update on image upload during project
  creation
* Removed unecessary console prints
* App now redirects to projects page on project creation
